src/client.rs

```diff
@@ -126,6 +136,48 @@ impl Client {
         Ok(response)
     }
 
+    /// Sends a batch of requests to the client.  The return vector holds the response
+    /// for the request at the corresponding index.  If no response was provided, it's [None].
+    ///
+    /// Note that the requests need to have valid IDs, so it is advised to create the requests
+    /// with [build_request].
+    pub fn send_batch(&self, requests: &[Request]) -> Result<Vec<Option<Response>>, Error> {
+        if requests.len() < 1 {
+            return Err(Error::EmptyBatch);
+        }
+
+        // If the request body is invalid JSON, the response is a single response object.
+        // We ignore this case since we are confident we are producing valid JSON.
+        let mut responses: Vec<Response> = self.send_raw(&requests)?;
+        if responses.len() > requests.len() {
+            return Err(Error::WrongBatchResponseSize);
+        }
+
+        // Match responses to the requests.
+        let mut results = Vec::with_capacity(requests.len());
+        for req in requests {
+            let mut matching = None;
+            for (i, resp) in responses.iter().enumerate() {
+                if resp.id == req.id {
+                    matching = Some(i);
+                    break;
+                }
+            }
+
+            match matching {
+                Some(m) => results.push(Some(responses.remove(m))),
+                None => results.push(None),
+            }
+        }
```

Comment:
Yikes :) this is O(n^2).

Can you instead use an algorithm where you
1. Loop over all responses, putting them in an id→Response `HashMap`,
2. To generate the returned array, loop over all requests, removing objects from the `HashMap` as you go?

Reply:
Hmm, yeah did this making the assumption that most servers would just order them correct anyway, making this O(n). But I guess we shouldn't make that assumption.

Reply:
Please review the updated code.